ccm_web/client/src/utils/enrollmentValidators.ts

```diff
@@ -0,0 +1,113 @@
+import { ClientEnrollmentType, isValidRole } from '../models/canvas'
+import { InvalidationType } from '../models/models'
+import { DuplicateIdentifierInRowsValidator, RowInvalidation } from '../utils/rowValidation'
+import {
+  emailSchema, firstNameSchema, lastNameSchema, loginIDSchema, validateString, ValidationResult
+} from '../utils/validation'
+
+export interface EnrollmentInvalidation extends RowInvalidation {}
+
+interface EnrollmentRowsValidator {
+  validate: (values: string[]) => EnrollmentInvalidation[]
+}
+
+const getMessage = (result: ValidationResult, fieldName: string): string => {
+  return result.messages.length > 0 ? result.messages[0] : `Value for ${fieldName} is invalid.`
+}
+
+export class EmailRowsValidator implements EnrollmentRowsValidator {
+  validate (emails: string[]): EnrollmentInvalidation[] {
+    const invalidations: EnrollmentInvalidation[] = []
+    emails.forEach((email, i) => {
+      const emailValidationResult = validateString(email, emailSchema)
+      if (!emailValidationResult.isValid) {
+        invalidations.push({
+          rowNumber: i + 2,
+          message: getMessage(emailValidationResult, 'email address'),
+          type: InvalidationType.Error
+        })
+      }
+    })
+    return invalidations
+  }
+}
+
+export class DuplicateEmailRowsValidator implements EnrollmentRowsValidator {
+  validate (emails: string[]): EnrollmentInvalidation[] {
+    const dupEmailValidator = new DuplicateIdentifierInRowsValidator('email address')
+    return dupEmailValidator.validate(emails)
+  }
+}
+
+export class LastNameRowsValidator implements EnrollmentRowsValidator {
+  validate (lastNames: string[]): EnrollmentInvalidation[] {
+    const invalidations: EnrollmentInvalidation[] = []
+    lastNames.forEach((lastName, i) => {
+      const lastNameValidationResult = validateString(lastName, lastNameSchema)
+      if (!lastNameValidationResult.isValid) {
+        invalidations.push({
+          rowNumber: i + 2,
+          message: getMessage(lastNameValidationResult, 'last name'),
+          type: InvalidationType.Error
+        })
+      }
+    })
+    return invalidations
+  }
+}
+
+export class FirstNameRowsValidator implements EnrollmentRowsValidator {
+  validate (firstNames: string[]): EnrollmentInvalidation[] {
+    const invalidations: EnrollmentInvalidation[] = []
+    firstNames.forEach((firstName, i) => {
+      const firstNameValidationResult = validateString(firstName, firstNameSchema)
+      if (!firstNameValidationResult.isValid) {
+        invalidations.push({
+          rowNumber: i + 2,
+          message: getMessage(firstNameValidationResult, 'first name'),
+          type: InvalidationType.Error
+        })
+      }
+    })
+    return invalidations
```

Comment:
This block is repeated for most/ all validation, could this be generalized?

Reply:
Yeah, I thought about this, let me give it a shot.

Reply:
How's this? https://github.com/tl-its-umich-edu/canvas-course-manager-next/pull/270/commits/7238bc99ebf59b82dfaa6fefb40aeb045d37368b